torch/distributed/distributed_c10d.py

```diff
@@ -1122,6 +1144,37 @@ def all_gather_multigpu(output_tensor_lists,
         work.wait()
 
 
+def all_gather_object(object, group=group.WORLD, async_op=False):
+    """
+    Performs all_gather on a picklable object.
+    """
+    if _rank_not_in_group(group):
+        return
+    _check_default_pg()
+
+    byte_tensor, size = _object_to_tensor(object)
+
+    group_size = torch.distributed.get_world_size(group=group)
+    object_size_list = [torch.LongTensor([0]) for _ in range(group_size)]
+
+    # collect the object sizes from all the tensors
+    all_gather(object_size_list, size, group=group)
+
+    # create the list of output tensors needed for all_gather
+    max_object_size = max(object_size_list)
+    tensor_list = [torch.ByteTensor(size=(max_object_size,)) for _ in range(group_size)]
+
+    if size != max_object_size:
+        # all tensors input into gather should be of the same size, so pad the local tensor if it is not the max size.
+        padding = torch.ByteTensor(size=(max_object_size - size,))
+        byte_tensor = torch.cat((byte_tensor, padding), dim=0)
```

Comment:
Does this mean we are creating 3 tensor storages here? original `byte_tensor`, `padding`, and the padded `byte_tensor`. I wonder if it would be faster if we directly allocate an empty tensor of the max size, and then copy the original `byte_tensor` to it.

Reply:
It looks like we can first allocate a tensor of `max_size`, and copy the serialized object tensor into it. However, we do need to know `max_size` first, which requires creating the serialized object tensor and getting its size via `numel()`.